{phys,lib} [foss/2024a] Yambo 5.3.0 single precision, DeviceXlib 0.9.0

nicspalla opened the pull request:
This pull request collects the recipes to compile Yambo version 5.3.0 in single precision with the foss/2024a toolchains.
Starting from version 5.3.0, Yambo requires the DeviceXlib library, developed by the same Yambo developers.

nicspalla committed:
Created new recipes for Yambo v5.3.0
Created recipes for library DeviceXlib (new dependency of Yambo)
Created new recipes for PETSc and SLEPc compiled in single precision

nicspalla committed:
changed moduleclass for DeviceXlib library

ocaisa commented on easybuild/easyconfigs/d/DeviceXlib/DeviceXlib-0.9.0-GCC-12.3.0-fb.eb:
```diff
@@ -0,0 +1,54 @@
+preconfigopts = ' && '.join([
+    'export CPP="cpp -P"',
+    'export FPP="cpp -P -traditional"',
+    'export F90FLAGS="-O3 -g -fallow-argument-mismatch -fopenmp  -cpp -fopenmp"',
+    'export CFLAGS="-O3"',
```
These are hard overwriting the values set by EasyBuild (which include additional optimisation flags), some of these you can get by setting the appropriate toolchainopts, e.g.,
```
toolchainopts = {'openmp': True, 'opt': True, 'debug': True}
```
would get you `-fopenmp`, `-g` and `-O3`. If you really need the others you can use
```
    'export F90FLAGS="-fallow-argument-mismatch $F90FLAGS"',
```

nicspalla replied:
solved as suggested using toolchainopts

ocaisa commented on easybuild/easyconfigs/d/DeviceXlib/DeviceXlib-0.8.6-GCC-13.3.0-fb.eb:
```diff
@@ -0,0 +1,54 @@
+
+name = 'DeviceXlib'
+version = '0.8.6'
+versionsuffix = '-fb'
```
What does the suffix mean? Is it just that it uses FlexiBLAS? If so I would drop it and use a toolchain that includes FlexiBLAS. Since you don't care about MPI here, this would be `gfbf`

nicspalla replied:
suffix removed

ocaisa commented on easybuild/easyconfigs/p/PETSc/PETSc-3.20.3-gompi-2023a-single-fb.eb:
```diff
@@ -0,0 +1,52 @@
+
+name = 'PETSc'
+version = '3.20.3'
+versionsuffix = '-single-fb'
```
What does the suffix mean? This is in a different toolchain to our other PETSc easyconfigs, why is that?

nicspalla replied:
file removed
see https://github.com/easybuilders/easybuild-easyblocks/pull/4139

ocaisa commented on easybuild/easyconfigs/p/PETSc/PETSc-3.20.3-gompi-2023a-single-fb.eb:
```diff
@@ -0,0 +1,52 @@
+    ('FlexiBLAS', '3.3.1'),
+]
+
+configopts = ' --LIBS="$LIBS -lrt" '
+configopts += '--with-scalar-type="complex" '
+configopts += '--with-precision="single" '
+configopts += '--with-blas-lapack-lib="$EBROOTFLEXIBLAS/lib/libflexiblas.a -lgfortran" '
```
I'm trying to understand what is special about this build that it is different to what we already ship (https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/p/PETSc/PETSc-3.20.3-foss-2023a.eb). The precision and scalar-type used here are non-standard in some way and not available in the other recipe?

ocaisa replied:
A chat with my AI friend tells me this may indeed be important for performance considerations, and that these choices are an unusual combination.

ocaisa replied:
You would somehow have to represent both of these in the `versionssuffix` to define a namespace for the other possibilties, perhaps
```
versionsuffix = '-single-complex'
```

nicspalla replied:
file removed
See https://github.com/easybuilders/easybuild-easyblocks/pull/4139

ocaisa commented on easybuild/easyconfigs/p/PETSc/PETSc-3.20.3-gompi-2023a-single-fb.eb:
```diff
@@ -0,0 +1,52 @@
+#      Author:    Jasper Grimm (UoY)
+##
+
+easyblock = 'ConfigureMake'
```
PETSc has its own easyblock with a lot of logic inside (https://github.com/easybuilders/easybuild-easyblocks/blob/develop/easybuild/easyblocks/p/petsc.py), why not use it?

ocaisa replied:
I now see in your original comment that for things to compile correctly you needed to avoid the easyblock. However, I would suspect that at least part of the reason for that need is that you are not using the full toolchain `foss` (which triggers the setting of a lot of environment variables).

nicspalla replied:
file removed
see PR https://github.com/easybuilders/easybuild-easyblocks/pull/4139

ocaisa commented on easybuild/easyconfigs/p/PETSc/PETSc-3.20.3-gompi-2023a-single-fb.eb:
```diff
@@ -0,0 +1,52 @@
+sanity_check_paths = {
+    'files': [],
+    'dirs': ['lib', 'lib64', 'include']
+}
```
This is a very weak sanity check, it is basically not checking anything

ocaisa commented on easybuild/easyconfigs/s/SLEPc/SLEPc-3.22.2-gompi-2024a-single-fb.eb:
```diff
@@ -0,0 +1,43 @@
+# Author: Nicola Spallanzani <nicola.spallanzani@cnr.it>
+##
+
+easyblock = 'ConfigureMake'
```
This has it's own easyblock, why not use it?

nicspalla replied:
file removed
see PR https://github.com/easybuilders/easybuild-easyblocks/pull/4139

ocaisa commented on easybuild/easyconfigs/s/SLEPc/SLEPc-3.22.2-gompi-2024a-single-fb.eb:
```diff
@@ -0,0 +1,43 @@
+easyblock = 'ConfigureMake'
+
+name = 'SLEPc'
+version = '3.22.2'
+versionsuffix = '-single-fb'
```
Is there something here that is not already provided by https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/s/SLEPc/SLEPc-3.23.2-foss-2024a.eb

nicspalla replied:
file removed

ocaisa commented on easybuild/easyconfigs/y/Yambo/Yambo-5.3.0-foss-2023a.eb:
```diff
@@ -0,0 +1,99 @@
+    {'Yambo-5.3_fix-timing.patch': 'e12b0da1038b7542222856c50dd483015b010678158d4f60eb0e2f5c0df60f13'},
+]
+dependencies = [
+    ('FFTW.MPI', '3.3.10'),
```
If you change the toolchain to foss (which it should be), this wouldn't be necessary

nicspalla replied:
file removed

ocaisa commented on easybuild/easyconfigs/y/Yambo/Yambo-5.3.0-foss-2023a.eb:
```diff
@@ -0,0 +1,99 @@
+    ('HDF5', '1.14.0'),
+    ('netCDF', '4.9.2'),
+    ('netCDF-Fortran', '4.6.1'),
+    ('FlexiBLAS', '3.3.1'),
+    ('ScaLAPACK', '2.2.0', '-fb'),
```
Part of `foss` toolchain

nicspalla replied:
file removed

ocaisa commented on easybuild/easyconfigs/y/Yambo/Yambo-5.3.0-foss-2023a.eb:
```diff
@@ -0,0 +1,99 @@
+    'export CPP="gcc -E -P"',
+    'export CPPFLAGS="-P"',
+    'export FPP="gfortran -E -P -cpp"',
+    'export FCFLAGS="-O3 -g -mtune=native -fno-lto -fopenmp"',
+    'export FFLAGS="-O3 -g -mtune=native -fno-lto -fopenmp"',
+    'export CFLAGS="-O2 -D_C_US -D_FORTRAN_US"',
```
You are hard overwriting the variables set by EasyBuild and I don't think you need to

nicspalla replied:
file removed

ocaisa reviewed (changes requested):
I'm curious about the options used for PetSC, I'd be open to including an suffix version of that if you think it is really necessary, but otherwise you should try to reuse what is already there.

There are cases where you don't use the appropriate toolchain for your easyconfigs, which is introducing additional complexity. Also, you are hard overwriting flags set by EasyBuild, which makes it harder for things like hooks to have expected impact.

ocaisa commented on easybuild/easyconfigs/s/SLEPc/SLEPc-3.20.1-gompi-2023a-single-fb.eb:
```diff
@@ -0,0 +1,43 @@
+
+#configopts = '--with-petsc-dir=$EBROOTPETSC '
+
+buildopts = 'PETSC_DIR=$EBROOTPETSC SLEPC_DIR=$PWD'
+installopts = 'PETSC_DIR=$EBROOTPETSC SLEPC_DIR=$PWD install'
```
These are interesting as we don't have them in the easyblock, do you know why they are needed?

nicspalla replied:
removed file

nicspalla committed:
Merge branch 'easybuilders:develop' into yambo-foss-single-precision

sorland2 committed:
DeviceXlib 0.9.0 easyconfig
- use toolchainopts for F90FLAGS
- remove -fb suffix for FlexiBLAS

sorland2 committed:
Yambo 5.3.0 easyconfigs
- use toolchainopts for FCFLAGS FFLAGS CFLAGS options
- DeviceXlib remove -fb suffix for FlexiBLAS

mmontag2 committed:
DeviceXlib-0.9.0-GCC-13.3.0.eb foss2024a
fix Moduleclass=lib

mmontag2 committed:
Yambo-5.3.0-foss-2023a.eb
-rm file Yambo-5.3.0-foss-2023a.eb from repo

nicspalla committed:
removed unnecessary files

boegel commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16

nicspalla committed:
removed PETSc and SLEPc dependencies from Yambo easyconfig (single precision)

nicspalla committed:
fix fail of test_style_conformance

nicspalla committed:
Merge branch 'easybuilders:develop' into yambo-foss-single-precision

nicspalla committed:
Merge branch 'easybuilders:develop' into yambo-foss-single-precision

nicspalla committed:
Yambo timing patch: added description

nicspalla committed:
fix checksum for patch file Yambo-5.3_fix-timing.patch

nicspalla committed:
Merge branch 'easybuilders:develop' into yambo-foss-single-precision

boegel commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16

boegel commented on easybuild/easyconfigs/d/DeviceXlib/DeviceXlib-0.9.0-GCC-13.3.0.eb:
```diff
@@ -0,0 +1,59 @@
+
+buildopts = 'all'
+
+sanity_check_paths = {
+    'files': ['lib/libdevXlib.a'],
+    'dirs': ['bin']
+}
```
It always make more sense to check for a couple of specific things, rather than just accepting a non-empty directory. And we might as well also run these `test_*` binaries as sanity check commands, while we're at it, since they're quick:

```suggestion
local_test_bins = ['buffer', 'linalg', 'malloc', 'mapping', 'memcpy_async', 'memcpy', 'memset', 'pinned']

sanity_check_paths = {
    'files': [f'bin/test_{b}.x' for b in local_test_bins] + ['lib/libdevXlib.a'],
    'dirs': [],
}

sanity_check_commands = [f'test_{b}.x' for b in local_test_bins]

boegel committed:
enhance sanity check for DeviceXlib

boegel commented:
> Test report by @boegelbot **FAILED** Build succeeded for 1 out of 2 (total: 50 secs) (2 easyconfigs in total) jsczen3c2.int.jsc-zen3.fz-juelich.de - Linux Rocky Linux 9.7, x86_64, AMD EPYC-Milan Processor (zen3), Python 3.9.25 See https://gist.github.com/boegelbot/79df75dc203c6c2e56ee7c3866a3b768 for a full test report.

There was an old copy of `iotk-y1.2.2.tar.gz` in place, which has a different checksum:
```
Checksum verification for /project/def-maintainers/boegelbot/sources/y/Yambo/iotk-y1.2.2.tar.gz using {'iotk-y1.2.2.tar.gz': '64af6a4b98f3b62fcec603e4e1b00ef994f95a0efa53ab6593ebcfe6de1739ef'} failed.
```

This was downloaded long ago (1 Dec 2023), I suspect the tarball being served from https://github.com/yambo-code/yambo-libraries/tree/master/external has somehow changed since then...

This old source tarball is not used anymore, so I've just moved it out of the way, new test report should come back successfully...

boegel commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 21 mins 27 secs) (2 easyconfigs in total)
node4412.skiddo.os - Linux RHEL 9.6, x86_64, AMD EPYC 9755 128-Core Processor (zen5), Python 3.9.21
See https://gist.github.com/boegel/d07b6882e4de3b2351a8a719f769a9ca for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 8 mins 43 secs) (2 easyconfigs in total)
aarch64-neoverse-v1-node1.int.aws-rocky88-202310.eessi.io - Linux Rocky Linux 8.10 (Green Obsidian), AArch64, ARM UNKNOWN (neoverse_v1), Python 3.13.4
See https://gist.github.com/boegel/1978c13c920f97b0e802d32db978a167 for a full test report.

edit: this test was on top of EESSI 2025.06

boegel commented on easybuild/easyconfigs/y/Yambo/Yambo-5.3.0-foss-2024a.eb:
```diff
@@ -0,0 +1,96 @@
+toolchain = {'name': 'foss', 'version': '2024a'}
+toolchainopts = {
+    'usempi': True, 'openmp': True,
+    'debug': True, 'opt': True, 'vectorize': None,
```
@nicspalla Overlooked this initially: can you clarify why `vectorize` is disabled?

Did you see trouble when it was enabled?

nicspalla replied:
I have to say that I don't remember. 
Removing `'vectorize': None` the compilation works on x86.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (total: 41 mins 38 secs) (1 easyconfigs in total)
node4235.shinx.os - Linux RHEL 9.6, x86_64, AMD EPYC 9654 96-Core Processor (zen4), Python 3.9.21
See https://gist.github.com/boegel/ba7a850dc130b2edc9bccbe1ae4c2875 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 34 mins 13 secs) (2 easyconfigs in total)
node3501.doduo.os - Linux RHEL 9.6, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.9.21
See https://gist.github.com/boegel/624972a0353ef96c3adca9262532a969 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 39 mins 45 secs) (2 easyconfigs in total)
node4005.donphan.os - Linux RHEL 9.6, x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), 1 x NVIDIA NVIDIA A2, 595.71.05, Python 3.9.21
See https://gist.github.com/boegel/567a5665ebe4d90b12813c7b2ad3d267 for a full test report.

nicspalla committed:
Removed vectorize: None in the Yambo recipe (see PR #23885)

nicspalla committed:
Merge branch 'easybuilders:develop' into yambo-foss-single-precision

boegel commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 32 mins 0 secs) (2 easyconfigs in total)
node3523.doduo.os - Linux RHEL 9.6, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.9.21
See https://gist.github.com/boegel/e3bc09ea92d340444e2ed9c61f8a1756 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 40 mins 7 secs) (2 easyconfigs in total)
node4008.donphan.os - Linux RHEL 9.6, x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), 1 x NVIDIA NVIDIA A2, 595.71.05, Python 3.9.21
See https://gist.github.com/boegel/c030efafa8dae3ba1df110184f05e7b0 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 7 mins 18 secs) (2 easyconfigs in total)
aarch64-neoverse-v1-node1.int.aws-rocky88-202310.eessi.io - Linux Rocky Linux 8.10 (Green Obsidian), AArch64, ARM UNKNOWN (neoverse_v1), Python 3.11.4
See https://gist.github.com/boegel/e60c61243fa1c9e963aed37a7a22ba00 for a full test report.

edit: this was on top of EESSI 2025.06

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (total: 1 hour 4 mins 40 secs) (2 easyconfigs in total)
node4231.shinx.os - Linux RHEL 9.6, x86_64, AMD EPYC 9654 96-Core Processor (zen4), Python 3.9.21
See https://gist.github.com/boegel/3db1ed0b1624d28fb2339ab58cb83807 for a full test report.

boegel reviewed (approved):
lgtm

ocaisa reviewed (approved):
LGTM

ocaisa commented:
Going in, thanks for all your patience @nicspalla !

nicspalla commented:
Thanks both of you for your help!
